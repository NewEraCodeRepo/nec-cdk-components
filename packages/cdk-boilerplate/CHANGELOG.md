# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.8.2](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.8.1...v1.8.2) (2020-01-10)

**Note:** Version bump only for package @NewEraCode/cdk-boilerplate





## [1.8.1](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.8.0...v1.8.1) (2019-10-23)

**Note:** Version bump only for package @NewEraCode/cdk-boilerplate





# 1.8.0 (2019-10-23)


### Bug Fixes

* merge fixes ([d45bfb0](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/d45bfb04a6cab25f96ac5e3ee895aa72cd223139))
* merge fixes ([4799bd6](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/4799bd6df82839b43e19d40be99f378352d05c74))


### Features

* update and add new mods ([0a21999](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/0a21999e9ade9130b52eacbabae9e2d635436987))





## [1.7.6](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.7.5...v1.7.6) (2019-10-22)

**Note:** Version bump only for package @NewEraCode/cdk-boilerplate





## [1.7.5](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.7.2...v1.7.5) (2019-10-22)


### Bug Fixes

* merge fixes ([4799bd6](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/4799bd6df82839b43e19d40be99f378352d05c74))
