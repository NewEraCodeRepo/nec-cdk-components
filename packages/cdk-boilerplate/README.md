# NewEraCode <Component Name>

## Description

An [AWS CDK](https://docs.aws.amazon.com/cdk/latest/guide/home.html) construct.

## Usage

### Install

```shell
$ npm i @NewEraCode/cdk-<component-name>
```
### Import

**NodeJS**
```javascript
const component = require('@NewEraCode/cdk-<component-name>');
```
**TypeScript / Babel**
```typescript
import { Component } from '@NewEraCode/cdk-<component-name>';
```

Example:

```typescript
import { Component } from '@NewEraCode/cdk-<component-name>';

// add exmaple of how to use component
```
