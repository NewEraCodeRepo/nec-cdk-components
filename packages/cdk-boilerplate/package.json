{
  "name": "@neweracode/cdk-boilerplate",
  "version": "1.8.2",
  "private": true,
  "description": "An AWS CDK construct",
  "author": "Your Name <you0@email.com>",
  "homepage": "https://github.com/NewEraCodeRepo/nec-cdk-components#readme",
  "license": "MIT",
  "main": "lib/cdk-boilerplate.js",
  "publishConfig": {
    "access": "restricted"
  },
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/NewEraCodeRepo/nec-cdk-components.git"
  },
  "scripts": {
    "tsc": "tsc",
    "watch": "tsc -w",
    "cdk": "cdk",
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "bugs": {
    "url": "https://github.com/NewEraCodeRepo/nec-cdk-components/issues"
  },
  "devDependencies": {
    "@aws-cdk/assert": "^1.26.0",
    "@babel/cli": "^7.8.4",
    "@babel/core": "^7.8.6",
    "@babel/preset-env": "^7.8.6",
    "@babel/preset-typescript": "^7.8.3",
    "@types/jest": "^25.1.3",
    "@types/node": "^13.7.7",
    "aws-cdk": "^1.26.0",
    "babel-core": "7.0.0-bridge.0",
    "ts-jest": "^25.2.1",
    "ts-node": "^8.6.2",
    "tslint": "^6.0.0",
    "tslint-config-prettier": "^1.18.0",
    "typescript": "^3.8.3"
  },
  "dependencies": {
    "@aws-cdk/aws-cloudtrail": "^1.26.0",
    "@aws-cdk/aws-s3": "^1.26.0",
    "@aws-cdk/core": "^1.26.0",
    "source-map-support": "^0.5.16"
  }
}
