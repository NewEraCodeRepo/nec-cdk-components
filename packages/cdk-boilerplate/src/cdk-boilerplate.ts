#!/usr/bin/env node
import * as cloudtrail from '@aws-cdk/aws-cloudtrail';
import { Bucket } from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';

export interface ICDKBoilerplateProps {
  bucketName: string;
}

/**
 * Boilerplate
 */
export class CDKBoilerplate extends Construct {
  public bucketWebsiteUrl: string;

  constructor(scope: Construct, id: string, props: ICDKBoilerplateProps) {
    super(scope, id);

    const bucket = new Bucket(this, 'CloudTrailBucket', {
      bucketName: props.bucketName,
    });
    // tslint:disable-next-line:no-unused-expression
    new cloudtrail.Trail(this, 'CloudTrail', {
      bucket,
      isMultiRegionTrail: true,
    });

    this.bucketWebsiteUrl = bucket.bucketWebsiteUrl;
  }
}
