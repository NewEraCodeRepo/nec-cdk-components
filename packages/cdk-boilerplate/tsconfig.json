{
  "extends": "../../base-tsconfig-ci.json",
  "compilerOptions": {
    "outDir": "./lib"
  },
  "exclude": ["cdk.out"],
  "include": ["./src"]
}
