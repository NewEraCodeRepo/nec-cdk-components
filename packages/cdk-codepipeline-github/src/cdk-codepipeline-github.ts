import { IStage, Pipeline } from '@aws-cdk/aws-codepipeline';
import { LambdaFunction } from '@aws-cdk/aws-events-targets';
import { PolicyStatement } from '@aws-cdk/aws-iam';
import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import { Construct, Duration, SecretValue } from '@aws-cdk/core';
import { join } from 'path';

export interface ICdkCodepipelineGithubProps {
  /**
   * The Codepipeline stage publishing status events
   */
  readonly stage?: IStage;

  /**
   * The Codepipeline publishing status events
   */
  readonly pipeline?: Pipeline;

  /**
   * The arn of a secret in the Secrets Manager with
   * a github token value.
   */
  readonly gitHubSecretArn: string;
}

/**
 * Publishes status events from a stage in Codepipeline
 * to a Github repo using webhooks.
 *
 * SAM template conversion by Julia Jacobs
 * {@see {@link https://github.com/jewelsjacobs}}
 *
 * SAM template created by Jens Eickmeyer
 * {@see {@link https://github.com/jenseickmeyer/github-commit-status-bot}}
 * {@see {@link https://scratchpad.blog/devops/howto/send-commit-status-from-codepipeline-to-github/}}
 *
 * @example <caption>Example usage of construct in pipeline.</caption>
 * // One possible way to get an existing secret's arn - using AwsCustomResource
 * // and AWS SecretsManager SDK
 * const getGithubSecret = new AwsCustomResource(this, 'GetGithubSecret', {
 *   onUpdate: {
 *      service: 'SecretsManager',
 *      action: 'describeSecret',
 *      parameters: {
 *         SecretId: 'TOKENS/PIPELINE/GithubToken' // ID of secret
 *      }
 *   }
 * });
 * // ARN is a property in the describeSecret SDK response
 * const gitHubSecretArn = `${getGithubSecret.getData("ARN")}` as string;
 *
 * // pipeline
 * const pipeline = new codepipeline.Pipeline(this, 'Pipeline', {
 *   restartExecutionOnUpdate: true
 * });
 *
 * // source stage
 * pipeline.addStage({
 *   stageName: 'Source',
 *   actions: [sourceAction]
 * });
 *
 * // build stage
 * const buildStage = pipeline.addStage({
 *   stageName: 'Build',
 *   actions: [buildAction]
 * });
 *
 * // Github status checks based on status of the build stage
 * new CdkCodepipelineGithub(this, 'GithubCodepipeline', {
 *   stage: buildStage,
 *   gitHubSecretArn
 * });
 *
 * // deploy stage
 * pipeline.addStage({
 *   stageName: 'Deploy',
 *   actions: [deployAction]
 * });
 *
 * // or check status of pipeline itself:
 * new CdkCodepipelineGithub(this, 'GithubCodepipeline', {
 *   pipeline,
 *   gitHubSecretArn
 * });
 */
export class CdkCodepipelineGithub extends Construct {
  constructor(scope: Construct, id: string, props: ICdkCodepipelineGithubProps) {
    super(scope, id);

    const { stage, pipeline, gitHubSecretArn } = props;

    const statusLambda = new Function(this, 'StatusLambda', {
      code: Code.asset(join(__dirname, '..', 'lambda_packages', 'package.zip')),
      environment: {
        ACCESS_TOKEN: SecretValue.secretsManager(gitHubSecretArn).toString(), // github token passed to Lambda
      },
      handler: 'index.handler',
      runtime: Runtime.NODEJS_10_X,
      timeout: Duration.seconds(300),
    });

    const lambdaTarget = new LambdaFunction(statusLambda);

    if (stage) {
      stage.onStateChange('StageOnStateChange', lambdaTarget, {
        eventPattern: {
          detail: {
            state: ['STARTED', 'SUCCEEDED', 'FAILED'],
          },
        },
      });
    }

    if (pipeline) {
      pipeline.onStateChange('StageOnStateChange', {
        eventPattern: {
          detail: {
            state: ['STARTED', 'SUCCEEDED', 'FAILED'],
          },
        },
        target: lambdaTarget,
      });
    }

    if (!stage && !pipeline) {
      throw Error('Choose a pipeline or stage to send event status messages from');
    }

    statusLambda.addToRolePolicy(
      new PolicyStatement({
        actions: ['*'],
        resources: ['*'],
      }),
    );
  }
}
