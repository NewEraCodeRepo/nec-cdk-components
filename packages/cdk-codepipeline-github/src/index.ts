export * from './cdk-codepipeline-github';
