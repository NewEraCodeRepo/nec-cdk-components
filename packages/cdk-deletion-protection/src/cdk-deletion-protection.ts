#!/usr/bin/env node
import { CfnDeletionPolicy, CfnResource, Construct } from '@aws-cdk/core';
import { AwsCustomResource } from '@aws-cdk/custom-resources';

export interface ICDKDeletionProtectionProps {
  resourceDeletionPolicy?: string; // "DELETE" | "RETAIN" | "SNAPSHOT"
  terminationProtection?: boolean;
}

/**
 * DeletionProtection
 *
 * Example usage:
 *
 * const deletionProtection = new CDKDeletionProtection(this, 'ProtectionPolicies', {
 *    resourceDeletionPolicy: "RETAIN",
 *    terminationProtection: true,
 * });
 *
 * // enable stack termination deletion protection:
 * deletionProtection.setStackTerminationProtection(this.stackName);
 *
 * // protect a Cognito User Pool from getting deleted on update:
 * const userPool = new UserPool(this, 'cdkUserPool', {...});
 * const cfnUserPool = userPool.node.defaultChild as CfnUserPool;
 * deletionProtection.setResourcePolicy(cfnUserPool);
 */
export class CDKDeletionProtection extends Construct {
  public setResourcePolicy: (cfnResource: CfnResource) => void;
  public setStackTerminationProtection: (stackName: string) => void;
  constructor(scope: Construct, id: string, props?: ICDKDeletionProtectionProps) {
    super(scope, id);

    this.setResourcePolicy = (cfnResource: CfnResource) => {
      if (props && props.hasOwnProperty('resourceDeletionPolicy')) {
        switch (props.resourceDeletionPolicy) {
          case 'DELETE':
            cfnResource.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.DELETE;
            cfnResource.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;
            break;
          case 'RETAIN':
            cfnResource.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.RETAIN;
            cfnResource.cfnOptions.deletionPolicy = CfnDeletionPolicy.RETAIN;
            break;
          case 'SNAPSHOT':
            cfnResource.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.SNAPSHOT;
            cfnResource.cfnOptions.deletionPolicy = CfnDeletionPolicy.SNAPSHOT;
            break;
          default:
            cfnResource.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.RETAIN;
            cfnResource.cfnOptions.deletionPolicy = CfnDeletionPolicy.RETAIN;
            break;
        }
      } else {
        throw new Error('Resource deletion policy required.');
      }
    };

    this.setStackTerminationProtection = (stackName: string) => {
      if (props && props.hasOwnProperty('terminationProtection')) {
        // Enable stack termination protection
        // using the AWS SDK CloudFormation updateTerminationProtection API:
        // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#updateTerminationProtection-property

        // tslint:disable-next-line:no-unused-expression
        new AwsCustomResource(this, 'UpdateTerminationProtection', {
          onUpdate: {
            action: 'updateTerminationProtection',
            parameters: {
              EnableTerminationProtection: props.terminationProtection,
              StackName: stackName,
            },
            service: 'CloudFormation',
          },
        });
      } else {
        throw Error('Termination protection option must be set to true or false.');
      }
    };
  }
}
