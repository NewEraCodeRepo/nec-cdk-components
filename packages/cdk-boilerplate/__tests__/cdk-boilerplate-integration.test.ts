import { App, Environment, Stack } from '@aws-cdk/core';
import 'jest-cdk-snapshot';
import { CDKBoilerplate, ICDKBoilerplateProps } from '../src/cdk-boilerplate';

describe('cdk-deletion-protection', () => {
  test('snapshot', () => {
    const env: Environment = {
      account: '123456789012',
      region: 'us-east-1',
    };

    const boilerplateProps: ICDKBoilerplateProps = {
      bucketName: 'test-bucket',
    };

    const app = new App();
    const stack = new Stack(app, 'Test', { env });
    // tslint:disable-next-line:no-unused-expression
    new CDKBoilerplate(stack, 'Boilerplate', boilerplateProps);

    expect(stack).toMatchCdkSnapshot();
  });
});
