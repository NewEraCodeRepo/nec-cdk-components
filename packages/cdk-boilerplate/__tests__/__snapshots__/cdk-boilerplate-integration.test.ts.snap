// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-deletion-protection snapshot 1`] = `
Object {
  "Resources": Object {
    "BoilerplateCloudTrail36F79572": Object {
      "DependsOn": Array [
        "BoilerplateCloudTrailBucketPolicy252E08F5",
      ],
      "Properties": Object {
        "EnableLogFileValidation": true,
        "EventSelectors": Array [],
        "IncludeGlobalServiceEvents": true,
        "IsLogging": true,
        "IsMultiRegionTrail": true,
        "S3BucketName": Object {
          "Ref": "BoilerplateCloudTrailBucket754399BE",
        },
      },
      "Type": "AWS::CloudTrail::Trail",
    },
    "BoilerplateCloudTrailBucket754399BE": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "test-bucket",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BoilerplateCloudTrailBucketPolicy252E08F5": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "BoilerplateCloudTrailBucket754399BE",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetBucketAcl",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "BoilerplateCloudTrailBucket754399BE",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "s3:PutObject",
              "Condition": Object {
                "StringEquals": Object {
                  "s3:x-amz-acl": "bucket-owner-full-control",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "BoilerplateCloudTrailBucket754399BE",
                        "Arn",
                      ],
                    },
                    "/AWSLogs/123456789012/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
